src/sentry/sentry_metrics/use_case_id_registry.py

```diff
@@ -0,0 +1,64 @@
+from __future__ import annotations
+
+from typing import Any, Iterator, Mapping, Optional, Set
+
+from sentry.sentry_metrics.configuration import UseCaseKey
+
+_HARDCODED_USE_CASES = {"PERFORMANCE": "performance", "RELEASE_HEALTH": "release-health"}
+
+_REGISTERED_USE_CASES: dict[str, str] = {}
+
+
+class _UseCaseID(type):
+    def __getattr__(self, attr: str) -> UseCaseID:
+        if attr not in _HARDCODED_USE_CASES and attr not in _REGISTERED_USE_CASES:
+            raise AttributeError(attr)
+
+        return UseCaseID(attr.lower())
+
+    def __iter__(self) -> Iterator[UseCaseID]:
+        return iter(
+            UseCaseID(value)
+            for value in {
+                **_HARDCODED_USE_CASES,
+                **_REGISTERED_USE_CASES,
+            }.values()
+        )
+
+
+class UseCaseID(metaclass=_UseCaseID):
```

Comment:
This whole class is pretty complex and unexpected for end users - can it not just be an enum?

Reply:
This is a class-based implementation of Enum. The main thing is that it lets us register use cases instead of just adding a new entry in an Enum.

Reply:
To elaborate a bit more, my idea behind not having just an Enum and instead supporting a `register` function was that it would kind of serve as an abstraction, which product engineers could call, to the underlying implementation. This way the underlying structure isn't being directly modified. The __iter__ also allows for iterating through the all of the existing `UseCaseID` without looking into what use cases are defined in an Enum. 

Maybe I overthought this though - do we think I should change this to something simpler?

Reply:
If you want end users to be able to register any string they like, then that's fine. But in that case I'm not sure I follow why you want the enum-like interface anyway if any string is allowed.

On the other hand, if you don't want users to register anything they like and we need to lock down the list of use cases, then I would switch to an real enum which actually prevents invalid use cases from being registered.

I'm not sure I followed the use case of having the combination of the 2 things - enum like interface but allowing for any random string to be registered there.

Reply:
I see what you mean about having a set of use cases that can be used (and not letting users register any string). In that case, do you think it would make sense to just have a regular Enum and we operate under the assumption that users of this interface would just have to add their own use case to that Enum as an entry before they can emit metrics for that use case?